Feature: Player leave/join island msgs

Obsidianninja11 opened the pull request:
### What
Adds an option to send leave/join msgs.
Configurable to work on public islands, while guesting, on your island, and only known players/known player's islands.

<details>
<summary>Images</summary>

![image](https://github.com/hannibal002/SkyHanni/assets/108832807/436900a2-42fb-4799-857f-683f2a1555b9)
</details>

## Changelog New Features
+ Option to send messages when players leave/join your island. - Obsidian

Obsidianninja11 committed:
Player leave/join island msgs

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/features/misc/LeaveJoinMsgsConfig.java:
```diff
@@ -0,0 +1,57 @@
+
+public class LeaveJoinMsgsConfig {
+
+    @Expose
+    @ConfigOption(
+        name = "Enabled",
+        desc = "Show a message in chat when a player leaves/joins your Private Island/Garden.\n§c§lMay be buggy in large hubs.")
+    @ConfigEditorBoolean
+    public boolean enabled = false;
```
This should have the feature toggle annotation

Obsidianninja11 replied:
wdym

CalMWolfs replied:
The @ FeatureToggle thing

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/features/misc/LeaveJoinMsgsConfig.java:
```diff
@@ -0,0 +1,57 @@
+    @ConfigEditorBoolean
+    public boolean leaveJoinColor = true;
+
+    @Expose
+    @ConfigOption(
+        name = "Only Known Players",
+        desc = "Only send a message if you know the person (they're in your friends, guild, or party).")
+    @ConfigEditorBoolean
+    public boolean onlyKnownPeople = false;
+
+    @Expose
+    @ConfigOption(
+        name = "Always on Your Island",
+        desc = "Bypass the Only Known Players toggle while on your Private Island/Garden.")
+    @ConfigEditorBoolean
+    public boolean alwaysOnYourIsland = false;
```
I feel both of these should be default true otherwise you will be spammed when turning on the feature with default settings

Obsidianninja11 replied:
Well if you just turn on Enabled, it'll only do it on your island
But yeah, probably

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+    private var updatedSinceWorldSwitch = false
+    private var onKnownIsland = false
+
+    private val rawPlayerPattern by RepoPattern.pattern(
+        "misc.islandleavejoinmsgs.rawplayers", "^§8\\[§\\w\\d+§8\\] (?<player>§\\w+).*$"
+    )
+    private val cleanPlayerPattern by RepoPattern.pattern(
+        "misc.islandleavejoinmsgs.cleanplayers", "^§8\\[§r§\\w\\d+§r§8\\] §r(?<player>§\\w+).*$"
+    )
+    private val offlinePlayerPattern by RepoPattern.pattern(
+        "misc.islandleavejoinmsgs.offlineplayers", "^§[0-9a-f]\\w+(?: §r§7\\(Offline [0-9dh+]+§r§7\\))?\$"
+    )
+    private val islandCategoryPattern by RepoPattern.pattern(
+        "misc.islandleavejoinmsgs.islandcategory", "^\\s+§r§b§lIsland$"
+    )
+    private val guestCategoryPattern by RepoPattern.pattern(
+        "misc.islandleavejoinmsgs.guestcategory", "^\\s+§r§5§lGuests §r§f\\(\\d+\\)$"
```
The pattern key and the pattern should be on new lines also should all be part of a pattern group

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt:
```diff
@@ -315,6 +316,11 @@ object Commands {
+            "shclearstoredplayers",
+            "Clears the stored players on the island."
+        ) { IslandLeaveJoinMsgs.clearPlayers() }
+
```
```suggestion
```
Remove new line

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+            )
+    }
+
+    private fun onPrivateIsland(guesting: Boolean = false) =
+        if (guesting) {
+            LorenzUtils.skyBlockIsland == IslandType.PRIVATE_ISLAND_GUEST || LorenzUtils.skyBlockIsland == IslandType.GARDEN_GUEST
+        } else LorenzUtils.skyBlockIsland == IslandType.PRIVATE_ISLAND || LorenzUtils.skyBlockIsland == IslandType.GARDEN
```
This should be a function in LocationUtils or IslandType instead

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+
+    private fun isPlayerKnown(player: String): Boolean {
+        return (
+            FriendAPI.getAllFriends().any { it.name.contains(player) } ||
```
I feel you should make a isFriend function instead of this, and then this function can be cleaned up

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+        onKnownIsland = false
+    }
+
+    private fun shouldSendMsg(player: String): Boolean {
+        val cleanPlayer = player.removeColor()
+        return (
+            LorenzUtils.getPlayerName() != cleanPlayer &&
+                if (config.onlyKnownPeople) isPlayerKnown(cleanPlayer) ||
+                    (config.alwaysOnYourIsland && onPrivateIsland()) ||
+                    (config.alwaysOnKnownIslands && onKnownIsland)
+                else true
+            )
+    }
```
```suggestion
        private fun shouldSendMsg(player: String): Boolean {
        val cleanPlayer = player.removeColor()
        
        if (LorenzUtils.getPlayerName() == cleanPlayer) return false
        if (!config.onlyKnownPeople) return true
        return isPlayerKnown(cleanPlayer) || (config.alwaysOnYourIsland && onPrivateIsland()) || (config.alwaysOnKnownIslands && onKnownIsland)
    }
```

Haven't tested this new function but this feels cleaner and more like existing code style

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+        var inIslandCategory = false
+        val islandOwners = mutableListOf<String>()
+
+        val joinMessage = " §" + if (config.leaveJoinColor) { "a" } else { "e" } + "joined §ethe island."
+        val leaveMessage = " §" + if (config.leaveJoinColor) { "c" } else { "e" } + "left §ethe island."
```
```suggestion
        val joinMessage = " §${if (config.leaveJoinColor) "a" else "e"}joined §ethe island."
        val leaveMessage = " §${if (config.leaveJoinColor) "c" else "e"}left §ethe island."
```

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+    private val config get() = SkyHanniMod.feature.misc.leaveJoinMsgs
+
+    private var players = mutableListOf<String>()
+    private var rawPlayersNew = mutableListOf<String>()
+    private var cleanPlayersNew = mutableListOf<String>()
```
These are only used inside of the `onTabListUpdate` function why are they here and not local to the function?

Obsidianninja11 replied:
uhhhhhh
idk lmao

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+            updatedSinceWorldSwitch = true
+        }
+
+        if (players.isEmpty()) return
```
Why not check this before the size?

CalMWolfs reviewed (changes requested):
Didnt test it but there is a lot of code cleanup to be done

Obsidianninja11 committed:
Code cleanup

Obsidianninja11 committed:
Fixed guesting detection

hannibal002 commented:
Is this a partial dupe to https://github.com/hannibal002/SkyHanni/pull/756?

Obsidianninja11 commented:
> Is this a partial dupe to #756?

Seems like it

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

Obsidianninja11 committed:
Fixes + code cleanup + known players options

Obsidianninja11 commented:
Should I make it send a different msg when you're on your private island/garden? Like: `[player] joined/left **your** island/**garden**.` instead of `[player] joined/left the island.`
It would allow people that are keeping others afk on their garden/island to set a chat notif for
Would be useful for people keeping others on their garden/island for social xp.

hannibal002 commented:
> Should I make it send a different msg when you're on your private island/garden? 

yes, this sounds useufl.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt:
```diff
@@ -369,6 +370,10 @@ object Commands {
             "Sets a test burrow waypoint at your location"
         ) { GriffinBurrowHelper.setTestBurrow(it) }
+        registerCommand(
+            "shclearstoredplayers",
```
I know finding a good name for this command is really hard, but the current name is confusing when not knowing its about island leave join messages. Too abstract/vague

Obsidianninja11 replied:
Do you have a suggestion?

Obsidianninja11 replied:
What if I improve the description?
"Clears the stored players on the island` for island leave/join messages`."

Obsidianninja11 replied:
maybe something like `/shclearislandplayers`?

hannibal002 replied:
Another thing, is this only intended for testing purposes? The list already gets cleared when the player itself switches worlds, right?
Maybe we can remove the command entirely? That way we don't need to find hard names XD

Obsidianninja11 replied:
> Another thing, is this only intended for testing purposes? 

pretty much

Maybe I could make a command to trigger the lorenzworldchangeevent?

hannibal002 replied:
Good alternative.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/misc/LeaveJoinMsgsConfig.java:
```diff
@@ -0,0 +1,115 @@
+    @Accordion
+    public KnownPlayersDetails knownPlayersDetails = new KnownPlayersDetails();
+
+    public static class KnownPlayersDetails {
```
move into own file

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:
```diff
@@ -93,6 +93,11 @@ public class MiscConfig {
     public TrackerConfig tracker = new TrackerConfig();
 
+    @Expose
+    @ConfigOption(name = "Leave/Join Messages", desc = "")
```
maybe include here in the name and description that it is about players on your island. otherwise its confusing with party members, friends, guild members, etc

Obsidianninja11 replied:
Like `Island Leave/Join Messages`?

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,145 @@
+        for ((index, player) in players.withIndex().reversed()) {
+            if (!playersNew.contains(player)) {
+                if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
+                players.removeAt(index)
```
why do we reverse this list?
cant we simply use a filter, send the chat message for every result, then remove the list from the players list?

Obsidianninja11 replied:
It's reversed because otherwise, removing indexes would break it
https://discord.com/channels/997079228510117908/1094190239532208228/1190852439641952296

There is probably a better way to do this than what i'm doing, but this seemed the simplest

hannibal002 replied:
Instead of reversing the list, you could create a copy of the list, iterate over it and remove the elements that way. No need to handle indexes at all then

Thunderblade73 replied:
isn't it just 
```
val iter = player.reversed().iterator()
while(iter.hasNext()){
val player = iter.next()
if (!playersNew.contains(player)) {
                if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
                iter.remove()
}
```

hannibal002 replied:
I would prefer:
```kt
for (player in players.filter { !playersNew.contains(it) }) {
    if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
    players.remove(player)
}

Obsidianninja11 replied:
> I would prefer:
> 
> ```kotlin
> for (player in players.filter { !playersNew.contains(it) }) {
>     if (shouldSendMsg(player)) ChatUtils.chat(player + leaveMessage)
>     players.remove(player)
> }
> ```

Oh that's way better than what I'm doing
I haven't used filter very much so I didn't think of that.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,145 @@
+        return isPlayerKnown(cleanPlayer) || (config.alwaysOnYourIsland && IslandType.onPrivateWorld()) || (config.alwaysOnKnownIslands && onKnownIsland)
+    }
+
+    private fun isPlayerKnown(player: String): Boolean {
```
this is duplicate to something in advanced tab list. write a common utils/api function instead

Obsidianninja11 replied:
https://discord.com/channels/997079228510117908/1094190239532208228/1213636854671474779

hannibal002 replied:
"Is player known (name) = get social score (name) > 1" (pseudo code)

Obsidianninja11 committed:
Different message on your island and garden

hannibal002 reviewed (changes requested):
Summary of still open requests:

- [ ] better remove player iteration
- [ ] no code duplication at isPlayerName
- [ ] delete or rename /shclearislandplayers

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,150 @@
+        return isPlayerKnown(cleanPlayer) || (config.alwaysOnYourIsland && IslandType.onPrivateWorld()) || (config.alwaysOnKnownIslands && onKnownIsland)
+    }
+
+    private fun isPlayerKnown(player: String): Boolean {
```
Swap the order of the two checks for every lne: first ask if the feature is enabled, then do the check.

Obsidianninja11 committed:
Code cleanup

Obsidianninja11 commented:
I deleted the clearstoredplayers command, and added simulateworldswitch
![image](https://github.com/hannibal002/SkyHanni/assets/108832807/cb5d3a20-022d-4a96-9347-9b6cfece49db)

I'm not sure if I should keep this though.

CalMWolfs commented:
> I deleted the clearstoredplayers command, and added simulateworldswitch
> 
> ![image](https://github.com/hannibal002/SkyHanni/assets/108832807/cb5d3a20-022d-4a96-9347-9b6cfece49db)
> 
> 
> 
> I'm not sure if I should keep this though.

I think it's fine to keep but why make a seperate function instead of just calling the event where you currently call the function

Obsidianninja11 commented:
> I think it's fine to keep but why make a seperate function instead of just calling the event where you currently call the function

Oh I didn't think of that lmao

Obsidianninja11 committed:
Made LorenzWorldChangeEvent be triggered in Commands.kt

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+    private val patternGroup = RepoPattern.group("misc.islandleavejoinmsgs")
+    private val rawPlayerPattern by patternGroup.pattern(
+        "rawplayers",
+        "§8\\[§[0-9a-f]\\d+§8\\] (?<player>§[0-9a-f]\\w+).*"
```
![image](https://github.com/hannibal002/SkyHanni/assets/24389977/04e9a4e5-168b-4b0e-91cf-70a186463496)

Obsidianninja11 replied:
Why?
That way seems kind of confusing if you edit the regex

hannibal002 reviewed (changes requested):
I cant get this feature to work at all with my second account in party/friends
The leave logic seems to work, not the join logic though.

Obsidianninja11 commented:
> I cant get this feature to work at all with my second account in party/friends


I don't think I actually tested known players after changing to using the social score function

wdym by leave works but join doesn't?
what island are you on?
ig that means that the function that makes it so you don't get spammed with join msgs when first joining a lobby isn't getting triggered?

Obsidianninja11 committed:
Fix

Obsidianninja11 commented:
Fix commit is for this bug:
https://discord.com/channels/997079228510117908/1094190239532208228/1215797101129371698

Maybe I should also add like a 0.5-1 sec cooldown after joining a new world?

hannibal002 reviewed (changes requested):
Standing in hub 1 with those settings:

<details>
<summary>Images</summary>

![image](https://github.com/hannibal002/SkyHanni/assets/24389977/5f67e4ac-5d9a-4815-81d3-2ccc2a47ab9c)

![image](https://github.com/hannibal002/SkyHanni/assets/24389977/1ef6546a-dabb-4ad5-8070-a1e46daf3c26)


</details>

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/misc/IslandLeaveJoinMsgs.kt:
```diff
@@ -0,0 +1,144 @@
+                if (players.contains(player)) return@matchMatcher
+                players.add(player)         // !onPrivateIslandGarden because a vanilla message gets sent
+                if (shouldSendMsg(player) && updatedSinceWorldSwitch && !onPrivateWorld) {
+                    ChatUtils.chat("$player$joinMessage")
```
Add option to remove the player name from chat message
```suggestion
                    ChatUtils.chat("${player.cleanPlayerName(displayName = true)}$joinMessage")
```

Obsidianninja11 replied:
~~wdym?~~
so the color?

Obsidianninja11 committed:
Support rank hider

Obsidianninja11 committed:
Fix debug stuff

Obsidianninja11 commented:
> Standing in hub 1 with those settings:
> 
> Images

I think that's from my fix 
I'll have to try something else

Obsidianninja11 committed:
Fix + adds cooldown after island switch

Obsidianninja11 commented:
Converting to draft bc I need to make it work with tab widgets and just better logic
more info: https://discord.com/channels/997079228510117908/1117409804382642296/1218742421471625247

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/config/features/misc/compacttablist/AdvancedPlayerListConfig.java
#	src/main/java/at/hannibal2/skyhanni/data/IslandType.kt
#	src/main/java/at/hannibal2/skyhanni/features/misc/compacttablist/AdvancedPlayerList.kt

Obsidianninja11 committed:
Fix

Obsidianninja11 committed:
Fix

Obsidianninja11 committed:
Fix (first try)

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/config/features/misc/compacttablist/AdvancedPlayerListConfig.java
#	src/main/java/at/hannibal2/skyhanni/features/misc/compacttablist/AdvancedPlayerList.kt

Obsidianninja11 committed:
Merge conflicts

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

Obsidianninja11 committed:
Fix

Obsidianninja11 committed:
Merge branch 'beta' into guest-leave-join-msgs

CalMWolfs commented:
@Obsidianninja11 are you planning on updating this pr or can I close it
